torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -301,43 +264,98 @@ static bool CheckEqual(const Expr* lhs, const Expr* rhs) {
+      const GPUMetaVarRewriter* metavars)
+      : thread_local_bufs_(thread_local_bufs) {
+    size_t DIMS = 3;
+
+    const std::vector<const Expr*>& block_extents =
+        metavars->gpu_block_extents();
+    const std::vector<const Var*>& block_vars = metavars->gpu_block_vars();
+    for (size_t i = 0; i < block_extents.size(); ++i) {
+      MetaVarExtent extent{block_extents[i], false};
+      if (extent.expr->isConstant() && immediateEquals(extent.expr, 1)) {
+        extent.trivial = true;
+      } else {
+        nontrivial_metavars_.insert(block_vars[i]);
+      }
+      metavars_[block_vars[i]] = extent;
+    }
+
+    const std::vector<const Expr*>& thread_extents =
+        metavars->gpu_thread_extents();
+    const std::vector<const Var*>& thread_vars = metavars->gpu_thread_vars();
+    for (size_t i = 0; i < thread_extents.size(); ++i) {
+      MetaVarExtent extent{thread_extents[i], false};
+      if (extent.expr->isConstant() && immediateEquals(extent.expr, 1)) {
+        extent.trivial = true;
+      } else {
+        nontrivial_metavars_.insert(thread_vars[i]);
+      }
+      metavars_[thread_vars[i]] = extent;
+    }
   }
-  return index_equal;
-}
 
-class AtomicAddFuser : public IRMutator {
   Stmt* mutate(const Store* v) override {
     const Buf* buf = v->buf();
-    const std::vector<const Expr*>& indices = v->indices();
-    const Expr* atomic_add_value = nullptr;
-    if (isAtomicAdd(v, &atomic_add_value)) {
-      return new AtomicAdd(buf, indices, atomic_add_value);
+    Store* orig = const_cast<Store*>(v); // NOLINT
+
+    // Thread locals never need to be atomic.
+    if (thread_local_bufs_.count(buf->base_handle()) != 0) {
+      return orig;
+    }
+
+    ScalarType dtype = v->value()->dtype().scalar_type();
+    if (dtype != ScalarType::Float && dtype != ScalarType::Double) {
+      return orig;
+    }
+    const Add* add_v = dynamic_cast<const Add*>(v->value());
+    if (!add_v) {
+      return orig;
+    }
+    const Load* load_v = dynamic_cast<const Load*>(add_v->lhs());
+    if (!load_v) {
+      return orig;
+    }
+    if (v->base_handle() != load_v->base_handle()) {
+      return orig;
+    }
+    if (v->indices().empty() && load_v->indices().empty()) {
+      return orig;
+    }
+    bool index_equal = CheckEqual(v->flat_index(), load_v->flat_index());
+    if (!index_equal) {
+      return orig;
+    }
+
+    // TODO: this checks that the metavars occur directly as an index, but this
+    // is pessimistic, blockIdx.x + 1 is fine too if there is no overlapping.
+    std::unordered_set<const Var*> vars_to_find = nontrivial_metavars_;
+    for (const Expr* e : v->indices()) {
+      if (const Var* v = dynamic_cast<const Var*>(e)) {
+        vars_to_find.erase(v);
+      }
+    }
+
+    if (vars_to_find.empty()) {
+      // All metavars accounted for.
+      return orig;
     }
-    return const_cast<Store*>(v); // NOLINT
+
+    return new AtomicAdd(buf, v->indices(), add_v->rhs());
```

Comment:
Great feature! This seems so fundamental that deserves its own directed tests. I would like to see a list of more explict expressions, and which will become AtomicAdd, vs others won't.

Reply:
You're right, this needs a lot more testing. The whole Cuda Codegen does. We'll have to come back to it.